### PR TITLE
feature: Webhook message edit & delete functionality

### PR DESCRIPTION
**Pull request by Desmont**
In December Discord added the functionality for webhooks to edit and delete webhook messages ( https://discord.com/developers/docs/change-log#december-15-2020 ) - this PR adds this functionality.

### CODE REVIEW

File: src/Discord.Net.Rest/API/Rest/EditWebhookMessageParams.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#pragma warning disable CS1591
+using Newtonsoft.Json;
+
+namespace Discord.API.Rest
+{
+    [JsonObject(MemberSerialization = MemberSerialization.OptIn)]
+    internal class EditWebhookMessageParams
```
Comment (SubZero0) [inline src/Discord.Net.Rest/API/Rest/EditWebhookMessageParams.cs]: Rename it to `ModifyWebhookMessageParams` and follow the same format as [ModifyMessageParams](https://github.com/discord-net/Discord.Net/blob/dev/src/Discord.Net.Rest/API/Rest/ModifyMessageParams.cs) (optional content, missing `get; set;`, no constructor).

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 2]
```diff
@@ -523,6 +523,44 @@ public async Task<Message> CreateWebhookMessageAsync(ulong webhookId, CreateWebh
             var ids = new BucketIds(webhookId: webhookId);
             return await SendJsonAsync<Message>("POST", () => $"webhooks/{webhookId}/{AuthToken}?wait=true", args, ids, clientBucket: ClientBucketType.SendEdit, options: options).ConfigureAwait(false);
         }
+
+        /// <exception cref="ArgumentOutOfRangeException">Message content is too long, length must be less or equal to <see cref="DiscordConfig.MaxMessageSize"/>.</exception>
+        /// <exception cref="InvalidOperationException">This operation may only be called with a <see cref="TokenType.Webhook"/> token.</exception>
+        public async Task EditWebhookMessageAsync(ulong webhookId, ulong messageId, EditWebhookMessageParams args, RequestOptions options = null)
```
Comment (SubZero0) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: Rename it to `ModifyWebhookMessageAsync`.

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 3]
```diff
@@ -523,6 +523,44 @@ public async Task<Message> CreateWebhookMessageAsync(ulong webhookId, CreateWebh
...
+        {
+            if (AuthTokenType != TokenType.Webhook)
+                throw new InvalidOperationException($"This operation may only be called with a {nameof(TokenType.Webhook)} token.");
+
+            Preconditions.NotNull(args, nameof(args));
+            Preconditions.NotEqual(webhookId, 0, nameof(webhookId));
+            Preconditions.NotEqual(messageId, 0, nameof(messageId));
+            if (!args.Embeds.IsSpecified || args.Embeds.Value == null || args.Embeds.Value.Length == 0)
```
Comment (SubZero0) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: Remove this, you don't know how the current webhook message looks like (if it has content, embeds, a file...).

File: src/Discord.Net.Rest/API/Rest/EditWebhookMessageParams.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        [JsonProperty("content")]
+        public string Content { get; }
+
```
Comment (SubZero0) [inline src/Discord.Net.Rest/API/Rest/EditWebhookMessageParams.cs]: Remove this empty line.

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 5]
```diff
@@ -523,6 +523,44 @@ public async Task<Message> CreateWebhookMessageAsync(ulong webhookId, CreateWebh
...
+                Preconditions.NotNullOrEmpty(args.Content, nameof(args.Content));
+
```
Comment (SubZero0) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs:537]: Missing a precondition to limit the amount of embeds to a maximum of 10 if it's being specified.

File: src/Discord.Net.Webhook/DiscordWebhookClient.cs [hunk 6]
```diff
@@ -91,6 +91,15 @@ private static API.DiscordRestApiClient CreateApiClient(DiscordRestConfig config
             string username = null, string avatarUrl = null, RequestOptions options = null, AllowedMentions allowedMentions = null)
             => WebhookClientHelper.SendMessageAsync(this, text, isTTS, embeds, username, avatarUrl, allowedMentions, options);
 
+        /// <summary> Edits a message posted using this webhook. </summary>
+        public Task EditMessageAsync(ulong messageId, string text = null, IEnumerable<Embed> embeds = null,
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/DiscordWebhookClient.cs]: Rename it to `ModifyMessageAsync`.

File: src/Discord.Net.Webhook/WebhookClientHelper.cs [hunk 7]
```diff
@@ -36,6 +36,21 @@ public static async Task<RestInternalWebhook> GetWebhookAsync(DiscordWebhookClie
             var model = await client.ApiClient.CreateWebhookMessageAsync(client.Webhook.Id, args, options: options).ConfigureAwait(false);
             return model.Id;
         }
+        public static async Task EditMessageAsync(DiscordWebhookClient client, ulong messageId, string text, IEnumerable<Embed> embeds,
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/WebhookClientHelper.cs]: Rename it to `ModifyMessageAsync`.

**Review by SubZero0 — CHANGES_REQUESTED** [review]

File: src/Discord.Net.Webhook/DiscordWebhookClient.cs [hunk 8]
```diff
@@ -91,6 +91,15 @@ private static API.DiscordRestApiClient CreateApiClient(DiscordRestConfig config
             string username = null, string avatarUrl = null, RequestOptions options = null, AllowedMentions allowedMentions = null)
             => WebhookClientHelper.SendMessageAsync(this, text, isTTS, embeds, username, avatarUrl, allowedMentions, options);
 
+        /// <summary> Edits a message posted using this webhook. </summary>
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/DiscordWebhookClient.cs]: Missing parameter documentation, follow the same pattern as https://github.com/discord-net/Discord.Net/blob/e3850e1e8f0f24652f22cc6322329495707ff67f/src/Discord.Net.Core/Entities/Messages/IUserMessage.cs#L38-L49

File: src/Discord.Net.Webhook/DiscordWebhookClient.cs [hunk 9]
```diff
@@ -91,6 +91,15 @@ private static API.DiscordRestApiClient CreateApiClient(DiscordRestConfig config
...
+        public Task EditMessageAsync(ulong messageId, string text = null, IEnumerable<Embed> embeds = null,
+            AllowedMentions allowedMentions = null, RequestOptions options = null)
+            => WebhookClientHelper.EditMessageAsync(this, messageId, text, embeds, allowedMentions, options);
+
+        /// <summary> Deletes a message posted using this webhook. </summary>
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/DiscordWebhookClient.cs]: Same as the other documentation.

**Review by SubZero0 — CHANGES_REQUESTED** [review]
Forgot the xml docs.

**Comment by Desmont** [comment]
Thanks for the feedback. I've made the changes that you requested.

File: src/Discord.Net.Rest/API/Rest/ModifyWebhookMessageParams.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+#pragma warning disable CS1591
+using Newtonsoft.Json;
+
+namespace Discord.API.Rest
+{
+    [JsonObject(MemberSerialization = MemberSerialization.OptIn)]
+    internal class ModifyWebhookMessageParams
+    {
+        [JsonProperty("content")]
+        public string Content { get; set; }
```
Comment (SubZero0) [inline src/Discord.Net.Rest/API/Rest/ModifyWebhookMessageParams.cs]: It's `Optional<string>`.

File: src/Discord.Net.Webhook/DiscordWebhookClient.cs [hunk 11]
```diff
@@ -91,6 +91,37 @@ private static API.DiscordRestApiClient CreateApiClient(DiscordRestConfig config
             string username = null, string avatarUrl = null, RequestOptions options = null, AllowedMentions allowedMentions = null)
             => WebhookClientHelper.SendMessageAsync(this, text, isTTS, embeds, username, avatarUrl, allowedMentions, options);
 
+        /// <summary> Modifies a message posted using this webhook. </summary>
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/DiscordWebhookClient.cs]: Let the text be on its own line, like the remarks.

File: src/Discord.Net.Webhook/DiscordWebhookClient.cs [hunk 12]
```diff
@@ -91,6 +91,37 @@ private static API.DiscordRestApiClient CreateApiClient(DiscordRestConfig config
...
+        /// </param>
+        /// <param name="options">The options to be used when sending the request.</param>
+        /// <returns>
+        ///     A task that represents the asynchronous modification operation.
+        /// </returns>
+        public Task ModifyMessageAsync(ulong messageId, string text = null, IEnumerable<Embed> embeds = null,
+            AllowedMentions allowedMentions = null, RequestOptions options = null)
+            => WebhookClientHelper.ModifyMessageAsync(this, messageId, text, embeds, allowedMentions, options);
+
+        /// <summary> Deletes a message posted using this webhook. </summary>
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/DiscordWebhookClient.cs]: Let the text be on its own line, like the remarks.

File: src/Discord.Net.Webhook/DiscordWebhookClient.cs [hunk 13]
```diff
@@ -91,6 +91,37 @@ private static API.DiscordRestApiClient CreateApiClient(DiscordRestConfig config
...
+        /// <param name="embeds">The <see cref="Discord.EmbedType.Rich" /> <see cref="Embed" /> enumeration to be sent.</param>
+        /// <param name="allowedMentions">
+        ///     Specifies if notifications are sent for mentioned users and roles in the message <paramref name="text"/>.
+        ///     If <c>null</c>, all mentioned roles and users will be notified.
+        /// </param>
+        /// <param name="options">The options to be used when sending the request.</param>
+        /// <returns>
+        ///     A task that represents the asynchronous modification operation.
+        /// </returns>
+        public Task ModifyMessageAsync(ulong messageId, string text = null, IEnumerable<Embed> embeds = null,
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/DiscordWebhookClient.cs]: Actually, this needs to be changed, follow the same pattern as modifying user messages.
See the following: https://github.com/discord-net/Discord.Net/blob/dev/src/Discord.Net.WebSocket/Entities/Messages/SocketUserMessage.cs#L166-L167

They need to optionally change what they want, a simple "null" isn't able to tell if they want to clear it or not send at all.

File: src/Discord.Net.Webhook/WebhookClientHelper.cs [hunk 14]
```diff
@@ -36,6 +36,21 @@ public static async Task<RestInternalWebhook> GetWebhookAsync(DiscordWebhookClie
             var model = await client.ApiClient.CreateWebhookMessageAsync(client.Webhook.Id, args, options: options).ConfigureAwait(false);
             return model.Id;
         }
+        public static async Task ModifyMessageAsync(DiscordWebhookClient client, ulong messageId, string text, IEnumerable<Embed> embeds,
```
Comment (SubZero0) [inline src/Discord.Net.Webhook/WebhookClientHelper.cs]: Also needs to follow the same pattern as modifying user message like stated above.

**Review by SubZero0 — CHANGES_REQUESTED** [review]

**Comment by MistressPlague** [comment]
This does not solve the issue I created, please investigate before proposing a pull will solve an issue.
I did not say webhook messages.

**Review by SubZero0 — APPROVED** [review]

**Pull request merged**